### PR TITLE
Fix to #5672 - Query :: Include with multiple navigations (including optional navigation) fails

### PR DESCRIPTION
**Pull request by maumar**
Problem was that our include logic for groupjoin was expecting outer and inner elements (on which the include was being performed) to be entities. However sometimes those elements would be TransparentIdentifiers (e.g. when the element comes from a result of SelectMany).

Fix is to also store (when necessary) the accessor from the outer/inner element to the entity that we want to include, extract the actual entity in runtime and apply include operations on that entity instead of the actual outer/inner element.

### CODE REVIEW

**Comment by maumar** [comment]
@anpete @divega

File: src/Microsoft.EntityFrameworkCore.Relational/Query/AsyncQueryMethodProvider.cs [hunk 1]
```diff
@@ -418,7 +418,15 @@ var outer
 
                         if (_groupJoinAsyncEnumerable._outerGroupJoinInclude != null)
                         {
-                            await _groupJoinAsyncEnumerable._outerGroupJoinInclude.IncludeAsync(outer, cancellationToken);
+                            var outerEntityAccessor = _groupJoinAsyncEnumerable._outerGroupJoinInclude.EntityAccessor as Func<TOuter, object>;
```
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/AsyncQueryMethodProvider.cs]: Can we only do this cast once?

File: src/Microsoft.EntityFrameworkCore.Relational/Query/AsyncQueryMethodProvider.cs [hunk 2]
```diff
@@ -385,8 +385,10 @@ private sealed class GroupJoinAsyncEnumerator : IAsyncEnumerator<TResult>
                 private IAsyncEnumerator<ValueBuffer> _sourceEnumerator;
                 private bool _hasNext;
                 private TOuter _nextOuter;
-                private AsyncGroupJoinInclude.AsyncGroupJoinIncludeContext outerGroupJoinIncludeContext;
-                private AsyncGroupJoinInclude.AsyncGroupJoinIncludeContext innerGroupJoinIncludeContext;
+                private AsyncGroupJoinInclude.AsyncGroupJoinIncludeContext _outerGroupJoinIncludeContext;
```
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/AsyncQueryMethodProvider.cs]: Oops! :smile:

File: src/Microsoft.EntityFrameworkCore.Relational/Query/AsyncQueryMethodProvider.cs [hunk 3]
```diff
@@ -418,9 +422,16 @@ var outer
 
                         _nextOuter = default(TOuter);
 
-                        if (outerGroupJoinIncludeContext != null)
+                        if (_outerGroupJoinIncludeContext != null)
                         {
-                            await outerGroupJoinIncludeContext.IncludeAsync(outer, cancellationToken);
+                            if (_outerEntityAccessor != null)
```
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/AsyncQueryMethodProvider.cs]: Use ternary

File: src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/Internal/IncludeExpressionVisitor.cs [hunk 4]
```diff
@@ -161,9 +171,18 @@ var groupJoinInclude
 
                 if (groupJoinInclude != null)
                 {
-                    var newArguments = methodCallExpression.Arguments.ToList();
+                    var groupJoinIncludeExpression = (Expression)Expression.Constant(groupJoinInclude);
```
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/Internal/IncludeExpressionVisitor.cs]: Why this cast?

**Comment by anpete** [comment]
:shipit: after R# cleanup.

**Pull request merged**